### ISSUE TITLE
[NORTHBOUND - MQTT] To provide an easy to use MQTT Client wrapper with QoS0

### ISSUE THREAD

**Issue by ralcaide** [Sinapse-Energia/EC-M2M#8]
In order to ease the use of the MQTT communications it is necessary to provide the following functions:

- MQTT_Connect:
Create a connect package using the code of #4 and send the package to the broker using the functions of #7 
return true or false

- MQTT_Disconnect:
Create a disconnect package using the code of #4 and send the package to the broker using the functions of #7 
return true or false

- MQTT_Publish_QoS0(msg, topic)
Create a publish package with the msg and topic using the code of #4 and send the package to the broker using the functions of #7. 
return true or false

- MQTT_Subscribe_QoS0(topic)
Create a subscribe package with the topic using the code of #4 and send the package to the broker using the functions of #7. 
return true or false

- MQTT_Receive_Message()
Read the buffer used in #7 in order to get the receive messages.
Return the message in order to process it

Basically is to create some wrapper functions in order to develop something like  https://github.com/eclipse/paho.mqtt.embedded-c/blob/master/MQTTPacket/samples/pub0sub1.c in the most easy way as possible.

This functions will be used by #1

**Comment by sinapseenergia** [comment]
Even if we are using Paho Library, it´s important to take a look over actuall Protocol Version:

http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html

Paho client simplifies all proccess but it doesn´t imlpement all client functionalities (at least in a basic configuration).
Features as MQTT broker answer threathment are useful for our implementation (To analyse disconections or view if a Client is banned)

Take a look over protocol messages Connect/ Subscribe and Publication functions taking care of the server responses in order to store and manage it.

**Comment by ralcaide** [comment]
Juanra has already worked on that. It is necessary to integrate the Francis work done in #7

**Comment by juanra-bascones** [comment]
I have been working in a kind of "proof of concept" with PAHO MQPacket file functions, and, finally I ended up with a much smaller and abstract interface over it.
Hopefully enough, the interface is pretty the same that te Rafa proposal.
I'm still making a little coding, and mostly testing, but the interface is very much the final one.

**Comment by juanra-bascones** [comment]
The MQTT  wrapper is built over Winsock libraries, compiled with MS Visual Studio and tested in a "neutral" PC, but I think is portable to Atollic without any change (so I hope)
The use of sockets API is done throug a quite simple interface, taken directlly from the transport.h file in the PAHO Github example.
I have tried to upload both files,  (the wrapper interface and the transport.h file), but it seems GitHub doesn´t allow update text files (or maybe I am quite thick..), so I just copy them in line:
The wrapper:
#ifndef MQTTAPI_H_
#define MQTTAPI_H_

extern	int		MqttConect	(const char	*host, unsigned int port, char *username, char *password);

extern	int		MqttSubscribe	(int handle, char *topic);

extern	char		*MqttGetMessage	(int handle, char *buf, int bufsize); // fgets style

extern	int		MqttPutMessage	(int handle, char *topic, char *message);

extern	int		MqttDisconnect	(int handle);

#endif

(as you can se, the main difference with Rafa proposal is the parameter list of some APIs)

The transport.h file is the paho untouched. I copy only the function prototypes:


int transport_sendPacketBuffer(int sock, unsigned char* buf, int buflen);
int transport_getdata(unsigned char* buf, int count);
int transport_getdatanb(void *sck, unsigned char* buf, int count);
int transport_open(const char* host, int port);
int transport_close(int sock);

**Comment by juanra-bascones** [comment]
As I understand, as far as Francis could implement a transport wrapper for the M95 GPRS driver with the very same interface, both transport layer will be perfectly interchangeables.

I trust to end testing the wrapper, and move the code to the M2MExample project in Atollic desktop this same morning, and pull down the server updates, in order to begin the integration as soon as possible.

**Comment by ralcaide** [comment]
I reopen the issue. Will be closed when the code is uploaded and integrated with the GPRS Wrapper

**Comment by juanra-bascones** [comment]
I have just uploaded the wrapper interface and implementation, and the transport interface.
There is a "subtle" issue about the existence of two, apparently redundant, functions
(getdata and getdanb..).
Let´s have it as is, so far.
// TODO :  Make some research about this, and try to consolidate in one function

**Comment by ralcaide** [comment]
Tested. It is working almost. It is necessary to design a reconnect mechanism.

Some comments to improve the Client added here: https://github.com/Sinapse-Energia/EC-M2M/commit/1bd7ae6b1f655735557997649bd15bd60676e3a1

**Comment by juanra-bascones** [comment]
In order to build the unique ClientID for the MQTT connection, I would like to know how can I get the device ID.
I understand that this ClientID (deviceId + random_number) is used to define the Suscription topic (where the input messages come from..) and also is part of the published messages (meassurement, etc...).
My purpose is to evaluate both of them, and write them to the context at the very beginning...thus, asking to the Context for the value whenever should be neeeded

**Comment by ralcaide** [comment]
Juanra:
The clientID is composed by DeviceID + Random_number. 
- With QoS0, the random_number should be generated each time the connection is performed in order to avoid problems with Broker
- With QoS1, the random number is not necessary or will be generated only during the first connection.

The subscription topics contains the DeviceID, not the ClientID. The DeviceID is always the same.

@francisjjp, could you explain how to get the DeviceID in a proper & clean way?

**Comment by juanra-bascones** [comment]
Ok.
As soon as we have some directions of how to get de DeviceID, we better use
it, instead of having it hardcoded...

But...where is the ClientID used for?

**Comment by ralcaide** [comment]
The ClientID is only used during the connection, in the Connect packet. The broker can not have to clients with the same ClientID.
At this moment is set in the MQTTConect function. Line 20:

data.clientID.cstring = "me1"

As you can see, this is not very well done. It should be improved

**Comment by ralcaide** [comment]
This version is stable. Alfa version.

**Comment by juanra-bascones** [comment]
I have added some new features, as general purpose utilities, but with special focus in diagnostics for the deafness effect. 
1) MqttPing() function, for sending PING message. Not really needed as far as we are sending traces every time, bus it is helpfull in knowing by sure if there is deafness (because receving or not the PINGRESP)
2) Refactoring of MqttConnect, splitting in two functions acting serielly, one for transport connection and other for broker connection, callables separatelly as needed
3) A new function Reconnect() managing the broker disconection and new connection KEEPING the transport connection underlying.
